{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "#The path to the biopython packages was not automatically added to anaconda. To allow import of the package, a manual path must be specified.\n",
    "import os\n",
    "#Absolute location: \n",
    "\n",
    "#os.environ['CONDA_PREFIX']=/home/wish1832/anaconda3\n",
    "#Code to append biopython location to path\n",
    "os.environ['PATH'] = os.environ['PATH']+';'+os.environ['CONDA_PREFIX']+\"/pkgs/biopython-1.78-py38h7b6447c_0/lib/python3.8/site-packages\""
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Based on analysis in R, Uniq2 (alignment[1]) contains the reference sequence. "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Scaffolding Code for Variant Calling"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Functions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 68,
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio import AlignIO as msa_in\n",
    "from Bio.Align import MultipleSeqAlignment as msa\n",
    "import Bio.SeqRecord\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "\n",
    "def initalize_file(outname):\n",
    "    \"\"\"\n",
    "    outname = the directory and prefix of the output file (string)\n",
    "        For example, setting outname \"mydir/mutant\" will create an outfile in the directory mydir with the name mutant_variants_raw.tsv.\n",
    "        Other reports generated by the output will also exist in the directory mydir and have the prefix \"mutant\".\n",
    "    \n",
    "    Given an output file prefix, initialize_files() will create one file (outname_variants_raw.tsv) \n",
    "    containing raw data on the substitutions, insertions, and deletions relative to the reference genome found in the sequence alignment. \n",
    "    \n",
    "    The function will print headers in each file in the order specified below:\n",
    "    \n",
    "    Field 1: Cluster_ID (e.g. clust48)  \n",
    "    Field 2: Number of sequences in cluster \n",
    "    Field 3: Type of variant (insertion, deletion, substitution)\n",
    "    Field 4: Mutation code (formatted based on the nomenclature indicated by the Human Genome Variation Society https://varnomen.hgvs.org/recommendations/protein/)\n",
    "    Field 5: Amino Acid(s) of Reference Strand \n",
    "        For insertions, this is the two amino acid residues the insertion ocurrs between.\n",
    "        For deletions, this will be the reference residue(s) deleted \n",
    "    Field 6: Variant Amino Acid Residue\n",
    "        For multi-codon insertions, this field will contain a string of multiple codons\n",
    "        For deletions, this will be blank\n",
    "    Field 7: Start position of variant on multiple sequence alignment\n",
    "        For insertions this will be the codon number on the multiple sequence alignment before the inserted codon(s)\n",
    "    Field 8: End position of variant on multiple sequence alignment \n",
    "        For insertions this will be the codon number on the multiple sequence alignment before the inserted codon(s)\n",
    "        For substitutions and single-base deletions this will be blank\n",
    "    Field 9: Start position of variant on reference strand \n",
    "        For insertions this will be the codon number on the reference strand before the inserted codon(s)\n",
    "    Field 10: End position of variant on reference strand\n",
    "        For insertions this will be the codon number on the reference strand before the inserted codon(s)\n",
    "        For substitutions and single-base deletions this will be blank\n",
    "    \"\"\"\n",
    "    #Create filename for raw output\n",
    "    raw_out = outname+\"_variants_raw.tsv\"\n",
    "    #Define header according to elements listed in docstring\n",
    "    header = [\"Cluster_ID\",\"Cluster_Size\",\"Type\",\"Code\",\"Ref Residue(s)\",\"Var Residue(s)\",\"AA_Start(MSA)\",\"AA_End(MSA)\",\"AA_Start(Ref)\",\"AA_End(Ref)\"]\n",
    "\n",
    "    #Print header in output file\n",
    "    with open(raw_out,\"w\") as outfile:\n",
    "        outfile.write(\"{}\\t{}\\t{}\\t{}\\t{}\\t{}\\t{}\\t{}\\t{}\\t{}\\n\".format(header[0],header[1],header[2],header[3],header[4],header[5],header[6],header[7],header[8],header[9]))\n",
    "\n",
    "    \"\"\"Code below will prevent overwriting in final version.\n",
    "        if not file:\n",
    "            with open(file,\"w\") as file\n",
    "        else\n",
    "        \n",
    "            print(\"Error: {} already exists. Overwriting prohibited.\".format(file))\"\"\"\n",
    "    #Return the filename for use in other functions\n",
    "    return raw_out\n",
    "\n",
    "def write_line(outfile,clustname,clustsize,var_type,code,ref_AA,var_AA,msa_start,ref_start,msa_end=\"-\",ref_end=\"-\"):\n",
    "    \"\"\"\n",
    "    Will write out to the variant .tsv values given the location of an output list with 6 Fields in the following format:\n",
    "    \n",
    "    Field 1: Cluster_ID (e.g. clust48)  \n",
    "    Field 2: Number of sequences in cluster \n",
    "    Field 3: Type of variant (insertion, deletion, substitution)\n",
    "    Field 4: Mutation code (formatted based on the nomenclature indicated by the Human Genome Variation Society https://varnomen.hgvs.org/recommendations/protein/)\n",
    "    Field 5: Amino Acid(s) of Reference Strand \n",
    "        For insertions, this is the two amino acid residues the insertion ocurrs between.\n",
    "        For deletions, this will be the reference residue(s) deleted \n",
    "    Field 6: Variant Amino Acid Residue\n",
    "        For multi-codon insertions, this field will contain multiple columns\n",
    "        For deletions, this will be blank\n",
    "    Field 7: Start position of variant on multiple sequence alignment\n",
    "        For insertions this will be the codon number on the multiple sequence alignment before the inserted codon(s)\n",
    "    Field 8: End position of variant on multiple sequence alignment \n",
    "        For insertions this will be the codon number on the multiple sequence alignment before the inserted codon(s)\n",
    "        For substitutions and single-base deletions this will be blank\n",
    "    Field 9: Start position of variant on reference strand \n",
    "        For insertions this will be the codon number on the reference strand before the inserted codon(s)\n",
    "    Field 10: End position of variant on reference strand\n",
    "        For insertions this will be the codon number on the reference strand before the inserted codon(s)\n",
    "        For substitutions and single-base deletions this will be blank\n",
    "    \"\"\"\n",
    "    #Write to output file\n",
    "    if outfile:\n",
    "        with open(outfile,\"a\") as file:\n",
    "            file.write(\"{}\\t{}\\t{}\\t{}\\t{}\\t{}\\t{}\\t{}\\t{}\\t{}\\n\".format(clustname,clustsize,var_type,code,ref_AA,var_AA,msa_start,msa_end,ref_start,ref_end))\n",
    "    \n",
    "    #Raise a FileNotFoundError if the output file has not been initialized\n",
    "    else:\n",
    "        raise FileNotFoundError(\"{} does not exist.\".format(outfile))        \n",
    "\n",
    "def reference_position(reference):\n",
    "    \"\"\"\n",
    "    For Variant Code Creation, the positions must correspond to the reference sequence, rather than the alignent.\n",
    "    \n",
    "    Given the reference genome, reference_position() will return a list of integers that stores the reference genome position corresponding to the msa position.\n",
    "    If there is no base in the reference genome at a given position in the msa (when a \"-\" is in the reference genome), a blank is stored in this list.\n",
    "    \"\"\"\n",
    "    ref_list=[]\n",
    "    pos=1\n",
    "    for i in reference:\n",
    "        if i != \"-\":\n",
    "            ref_list.append(pos)\n",
    "            pos += 1\n",
    "        else:\n",
    "            ref_list.append(\"-\")\n",
    "    return ref_list        \n",
    "        \n",
    "def msa_iterator1(alignment,raw_out,reference,ref_index,outfile):\n",
    "    \"\"\"\n",
    "    alignment=a biopython multiple sequence alignment object\n",
    "    raw_out=filename for outputting raw data\n",
    "    reference=Reference strand to which sequences are compared (biopython RefSeq object)\n",
    "    ref_list=A list relating an amino acid position on a multiple sequence alingment to its corresponding position on the reference strand\n",
    "    i_begin=Amino acid position on multiple sequence alignment at which to begin iteration (allows for skipping over multi-base insertions and deletions)\n",
    "    j_begin=Alignment object at which to begin\n",
    "    \n",
    "    \n",
    "    msa_iterator() will iterate through the multiple sequence alignment object and compare bases to the reference genome. \n",
    "    Insertions, deletions, and substitutions will be detected and handled by separate functions based on type. \n",
    "    \n",
    "    i is the amino acid position, j is the sequence position in the MSA (origional iterated through AA position first)\n",
    "    \"\"\"\n",
    "    #Every sequence in the alignment will be iteratively compared to the reference strand\n",
    "    #Multi-base deletions and insertions require a loop in which iterables can be edited\n",
    "    #A while loop will be used instead of a for loop to iterate through all i codons in j genomes.\n",
    "    \n",
    "    #Begin with i and j equal to zero\n",
    "    i,j=0,0\n",
    "    \n",
    "    #While loop will run until the last i position of the last j strain is tested.\n",
    "    while not (j==len(alignment)-1 and i==len(reference)):\n",
    "        #When all i codons in a genome have been iterated through, the genome index is advanced by one and the i is reset to zero.\n",
    "        if i==len(reference):\n",
    "            j+=1\n",
    "            i=0\n",
    "        else:\n",
    "            #For every codon, test the following:\n",
    "            print(\"testing j={}, i={}\".format(j,i))\n",
    "            #Conditional to detect a substitution\n",
    "            if reference[i] != alignment[j,i] and reference[i]!=\"-\" and alignment[j,i]!=\"-\":\n",
    "                print(\"AA substitution: position {0} in cluster {3}: {1}{0}{2}\".format(i+1,reference[i],alignment[j,i],j+1))\n",
    "                substitution_case(i,j,alignment,reference,ref_index,outfile)\n",
    "                #Move on to next codon\n",
    "                i+=1\n",
    "\n",
    "            #Conditional to detect a deletion\n",
    "            elif (reference[i] != alignment[j,i]) and alignment[j,i]==\"-\":\n",
    "                print(\"AA deletion: position {0} in cluster {3}: {1}{0}{2}\".format(i+1,reference[i],alignment[j,i],j+1))\n",
    "                print(\"j={}, i={}\".format(j,i))\n",
    "                #Call the deletion_case function, which will check for a multi-base deletion and return the value i to resume iteration from \n",
    "                i=deletion_case(j,i,alignment,reference,ref_index,outfile)\n",
    "\n",
    "            #Conditional to detect an insertion\n",
    "            elif reference[i] != alignment[j,i] and reference[i]==\"-\":\n",
    "                print(\"AA insertion: position {0} in cluster {3}: {1}{0}{2}\".format(i+1,reference[i],alignment[j,i],j+1))\n",
    "                i=insertion_case(j,i,alignment,reference,ref_index,outfile)\n",
    "                #insertion_case will set i to the next codon automatically\n",
    "            \n",
    "            #If there is no variant, increase i by one to test the next codon.\n",
    "            elif reference[i]==alignment[j,i]:\n",
    "                i+=1\n",
    "            \n",
    "            #Test for unanticipated scenarios\n",
    "            else:\n",
    "                print(\"A codon comparison has been found that does not meet any conditionals in the msa_iterator() loop.\")\n",
    "            \n",
    "    print(\"The while loop has been broken.\")\n",
    "        \n",
    "def substitution_case(i,j,alignment,reference,ref_index,outfile):\n",
    "    \"\"\"\n",
    "    Will activate when a substution is detected during iteration through the multiple sequence alignment. \n",
    "    \"\"\"\n",
    "    #Store position of variant in MSA (codon position is one more than the index)\n",
    "    msa_pos=i+1\n",
    "    #Store position of variant in reference genome\n",
    "    ref_pos=ref_index[i]\n",
    "    #Store codon in position i of reference genome\n",
    "    ref_AA=reference[i]\n",
    "    #Store the variant codon\n",
    "    var_AA=alignment[j,i] #Retrieves the codon from the sequence at position j, at position i\n",
    "    #Store the name of the cluster associated with the variant sequence (sequence index j)\n",
    "    #The cluster name is in the id of the RefSeq object, before the \";\".\n",
    "    clustname=alignment[j].id.split(sep=\";\")[0]\n",
    "    #Retrieve size of cluster associated with the variant sequence j\n",
    "    #The cluster size is printed betweeen \";\" and \"=\" in id\n",
    "    clustsize=alignment[j].id.split(sep=\"=\")[1].split(sep=\";\")[0]\n",
    "    #Store the substitution code\n",
    "    #Based on the sequence variant nomenclature suggested by the Human Genome Variation Society,\n",
    "    #The code format is <Reference_Amino_Acid><Codon_Position(Reference Genome)><Variant_Amino_Acid>\n",
    "    #Source (https://varnomen.hgvs.org/recommendations/protein/variant/substitution/)\n",
    "    code=\"{}{}{}\".format(ref_AA,ref_pos,var_AA)\n",
    "    \n",
    "    #Write above information to file\n",
    "    write_line(outfile,clustname,clustsize,\"sub\",code,ref_AA,var_AA,msa_pos,ref_pos,\"-\",\"-\")\n",
    "    \n",
    "    #Once information is stored, return to iteration at the next base (returns to the For loop in the parent function)\n",
    "\n",
    "def deletion_case(j,i,alignment,reference,ref_index,outfile):\n",
    "    \"\"\"\n",
    "    Will activate when a deletion is detected during iteration through the multiple sequence alignment. \n",
    "    \n",
    "    Since deletions may span two or more bases and nomenclature counts multi-base deletions as a single event, \n",
    "    iteration through the multiple sequence alignment in the parent msa_iterator() function must be skipped ahead\n",
    "    to the end of the detion region to avoid a multiple-codon deletion being read as multiple instances of a single-base deletion.\n",
    "    After recording a deletion, msa_iterator() is restarted at the position i corresponding to the codon after the end of the deletion. \n",
    "    \"\"\"\n",
    "    #Raise an error if the function is triggered by a codon that is not deleted\n",
    "    if not (alignment[j,i]==\"-\" and reference[i]!=\"-\"):\n",
    "        raise ValueError(\"Deletion case called on a codon region without a deletion\")\n",
    "    #Or if the deletion case is not called at the beginning of a deletion region\n",
    "    #Code detects a deletion in the codon one index before the current i it is called on\n",
    "    elif i!=0 and ((reference[i-1]!=alignment[j,i-1]) and alignment[j,i-1]==\"-\"): \n",
    "        raise ValueError(\"Deletion case called in the middle of a multi-base deletion\")\n",
    "\n",
    "    #Record the position at the beginning of a deletion\n",
    "    i_start=i   \n",
    "   \n",
    "    #Step 1: Iterate forward until extant codons are found on both the reference strand and genome j.\n",
    "    while True:\n",
    "        print(\"Testing i={}\".format(i))\n",
    "        #If a deletion case is detected, or if both the reference genome and genome j have a blank at index i, try the next codon.\n",
    "        if ((alignment[j,i]==\"-\" and reference[i]!=\"-\") or (alignment[j,i]==\"-\" and reference[i]==\"-\")):\n",
    "            print(\"Skipping blank at position {}\".format(i+1))\n",
    "            #If the base currently being tested at the end of the genome, a special case is triggered \n",
    "            if i==len(reference)-1:\n",
    "                print(\"Special case: deletion extends to end of genome.\")\n",
    "                #Set i_Cflank as being non-existent (affects which index is returned to msa_iterator() at end of function)\n",
    "                i_Cflank=None\n",
    "                break #Break the while loop and continue to step 2.\n",
    "            else:\n",
    "                i+=1 #Increase index by one and continue to test the next codon \n",
    "        \n",
    "        #When the C-flanking extant codon is found, record the index of the flank.\n",
    "        elif (alignment[j,i]!=\"-\" and reference[i]!=\"-\"):\n",
    "            print(\"Found C-terminal flanking codon at index {}.\".format(i))\n",
    "            #The index of the C-flanking codon will be returned to msa_iterator() as the position to resume variant checking.\n",
    "            i_Cflank=i\n",
    "            #Break the while loop and continue to step 2.\n",
    "            break \n",
    "            \n",
    "        #Special case: it is possible for an insertion to ocurr with a deletion (delins event).\n",
    "        elif alignment[j,i]!=\"-\" and reference[i]==\"-\":\n",
    "            print(\"Found a delins event.\")\n",
    "            #Run the delins case and store the index of the C-flanking codon to the event\n",
    "            i=delins_case(j,i_start,alignment,reference,ref_index,outfile)\n",
    "            #msa_iterator() will continue testing for variants at the index i\n",
    "            return i\n",
    "\n",
    "    #Step 2: Iterate backwards over any blanks that may exist until the end of the deletion region is reached.\n",
    "    print(\"Iterating backwards to find i_end.\")\n",
    "    print(\"Range of iterable\",range((i_Cflank if i_Cflank else len(reference)-1),i_start-1,-1),sep=\";\")\n",
    "    #Iteration will begin at the C-flanking codon (or the end of the genome) and will include the start of the deletion region.\n",
    "    for i in range((i_Cflank if i_Cflank else len(reference)-1),i_start-1,-1):\n",
    "        print(\"Testing i={}.\".format(i))\n",
    "        #If a blank is discovered at the position, continue to previous codon.\n",
    "        if (reference[i]==\"-\" and alignment[j,i]==\"-\"):\n",
    "            print(\"Skipping backwards over blank.\")\n",
    "        \n",
    "        #If a deletion is found, stop iteration and record current position as i_end (end of deletion region).\n",
    "        elif (reference[i]!=\"-\" and alignment[j,i]==\"-\"):\n",
    "            print(\"Discovered end of indel region. Recording index.\")\n",
    "            i_end=i\n",
    "            print(\"i_end=\",i_end,sep=\" \")\n",
    "            #Break the for loop and continue to step 3.\n",
    "            break\n",
    "    \n",
    "    #Step 3: Record deletion information using i_start and i_end.  \n",
    "    print(\"recording deletion info.\")\n",
    "    record_deletion_info(j,i_start,i_end,alignment,reference,ref_index,outfile)\n",
    "    \n",
    "    #Return the index of the C-flanking codon to msa_iterator() if it exists. \n",
    "    #For a deletion at the end of the genome, return i=len(reference), which will cause msa_iterator() to skip to the next line.\n",
    "    return (i_Cflank if i_Cflank else len(reference))\n",
    "        \n",
    "def insertion_case(j,i,alignment,reference,ref_index,outfile):\n",
    "    \"\"\"\n",
    "    insertion_case() will gather the following information regarding the insertion in 5 steps:\n",
    "    \n",
    "    instype=Either \"ins\", \"N_ext\", or \"C_ext\". instype will tell the function record_insertion_info() which\n",
    "        type of insertion case to record. Insertions at the 5' end of the transcriptome are N-terminal extentions,\n",
    "        and must be recorded differently from standard insertions. The same is true of C-terminal extentions, which\n",
    "        are insertions at the 3' end of the transcriptome.\n",
    "        \n",
    "        If N-terminal or C-terminal extentions have not been discovered, instype will be set to \"ins\".\n",
    "        \n",
    "    i_start=The index of the start of the insertion region (first location where reference is blank and the variant genome has a codon)\n",
    "    \n",
    "    i_Nflank= The index of the flanking N-terminal codon to the insertion in the reference strand (required for the code).\n",
    "        For N-terminal extentions, this will be equal to \"None\".\n",
    "    \n",
    "    i_end=The index of the end of the insertion region (last location where the reference is blank and the variant genome has a codon)\n",
    "    \n",
    "    i_Cflank=The index of the flanking C-terminal codon to the insertion in the reference strand (required for the insertion code).\n",
    "        For C-terminal extentions, this will be equal to \"None\".\n",
    "\n",
    "    \"\"\"\n",
    "    instype=None\n",
    "    \n",
    "    #Raise an error if the function is triggered by a codon that is not inserted\n",
    "    if not (alignment[j,i]!=\"-\" and reference[i]==\"-\"):\n",
    "        raise ValueError(\"Deletion case called on a codon region without a insertion\")\n",
    "    #Or if the insertion case is not called at the beginning of a deletion region\n",
    "    #Code detects a insertion in the codon one index before the current i it is called on\n",
    "    elif i!=0 and ((reference[i-1]!=alignment[j,i-1]) and reference[i-1]==\"-\"): \n",
    "        raise ValueError(\"Deletion case called in the middle of a multi-base insertion\")\n",
    "    \n",
    "    #Step 1: Record the index at beginning of insertion\n",
    "    i_start=i\n",
    "    \n",
    "    #Begin iteration\n",
    "    #Step 2: It is not necessarily true that the flanking reference codon is one index before the insertion start.\n",
    "    #A for loop will iterate backwards until an extant reference codon is found. \n",
    "    #If the insertion region is discovered in the first codon position, a special case is triggered.\n",
    "    if i!=0:\n",
    "        for idx in range(i-1,-1,-1): #Iterates backwards from the start of the insertion region, ending at 0\n",
    "            #idx used instead of i to prevent overwriting of i to be passed to be msa_iterator\n",
    "            if reference[idx]==\"-\":\n",
    "                print(\"Iterating backward over blank.\")\n",
    "                #Special case, if an N-terminal flanking codon has not been discovered when the beginning of the genome\n",
    "                #is reached, this is an N-terminal extention.\n",
    "                if idx==0:\n",
    "                    print(\"Insertion happens before start of reference sequence. This is an N-terminal extention.\")\n",
    "                    i_Nflank=None\n",
    "                    break #Break the loop and continue to step 3.\n",
    "            \n",
    "            #If the N-terminal flanking codon is found, store the index of the codon.\n",
    "            elif reference[idx]!=\"-\":\n",
    "                print(\"Found N-terminal flanking codon\")\n",
    "                i_Nflank=idx\n",
    "                break #Break the loop and continue to step 3.\n",
    "    \n",
    "    #Step 2: Special case of insertion at beginning of region \n",
    "    #Called when an insertion is found starting at i=0 (impossible for a N-terminal flank to exist)\n",
    "    else:\n",
    "        print(\"Special case: N-extention triggered.\")\n",
    "        i_Nflank=None\n",
    "        \n",
    "    #Step 3: Iterate forward until the first position of an extant codon on both the reference and the variant strand \n",
    "    #Will yield i_Cflank \n",
    "    print(\"Beginning forward iteration to find i_Cflank.\")\n",
    "    while True: \n",
    "        print(\"Testing j={}, i={}\".format(j,i))\n",
    "    \n",
    "        #Check for an insertion, or a blank on both the reference and the genome j, at the codon position. If so, continue on\n",
    "        if (reference[i]==\"-\" and alignment[j,i]!=\"-\") or (reference[i]==\"-\" and alignment[j,i]==\"-\"):\n",
    "            #Special case: end of the genome is reached and contains a blank (will also be triggered if the \n",
    "            #insertion begins at the end of the variant genome). This is a C-terminal extension.\n",
    "            if i==len(reference)-1:\n",
    "                print(\"End of genome reached. C-extension special case triggered\")\n",
    "                i_Cflank=None\n",
    "                break #Terminate the loop and continue to step 4  \n",
    "            \n",
    "            #If not at the end of the genome, increase the index by one and test the next codon.\n",
    "            else:\n",
    "                print(\"Blank at position {}\".format(i+1))\n",
    "                i+=1\n",
    "\n",
    "        #Check for a deletion at the codon. If so, switch to the indel case\n",
    "        elif reference[i]!=\"-\" and alignment[j,i]==\"-\":\n",
    "            print(\"Indel case found.\")\n",
    "            #i_start will be passed to the indel function\n",
    "            i=delins_case(j,i_start,alignment,reference,ref_index,outfile)\n",
    "            return i #After delins case is ran, return the position of the C-terminal flank if it exists to the msa_iterator() function\n",
    "\n",
    "        #Check for extant codons in both reference and genome j at the position\n",
    "        elif reference[i]!=\"-\" and alignment[j,i]!=\"-\":\n",
    "            #The position of the C-terminal flanking codon will be stored.\n",
    "            print(\"Reached C-flanking codon.\")\n",
    "            i_Cflank=i\n",
    "            break #Terminate the loop and continue to step 4.\n",
    "    \n",
    "    #Step 4: Iterate backward over any blank codons on both the reference and variant strands until a codon on the variant strand is reached\n",
    "    #Iteration will begin one codon before the C-terminal flank if it exists, otherwise iteration begins at the end of the genome.\n",
    "    print(\"Iterating backwards to discover the end of the insertion region.\")\n",
    "    for idx in range((i_Cflank-1 if i_Cflank else len(reference)-1),(i_Nflank if i_Nflank else -1),-1): #Iterates backwards to i_Nflank (excluding i_Nflank) if it exists, otherwise iterates to zero(inclusive).\n",
    "        print(\"Testing i={}\".format(idx))\n",
    "        #if a blank is found, continue backwards iteration \n",
    "        if alignment[j,idx]==\"-\":\n",
    "            print(\"Skipping backwards over blank.\")\n",
    "            continue\n",
    "        \n",
    "        #The first codon found in the alignment will mark the end of the inserted sequence in the multiple sequence alignment.\n",
    "        elif alignment[j,idx]!=\"-\":\n",
    "            print(\"The insertion region ends at position i={}. The ending codon is {}.\".format(idx,alignment[j,idx]))\n",
    "            i_end=idx\n",
    "            break #Break the loop and continue to step 5.\n",
    "    \n",
    "    #Step 5: determine type of insertion based on the presence/absence of N-terminal and C-terminal flanking codons\n",
    "    if i_Cflank and i_Nflank:\n",
    "        print(\"Insertion has both C- and N- terminal flanks.\")\n",
    "        instype=\"ins\"\n",
    "    elif i_Cflank and not i_Nflank:\n",
    "        print(\"Insertion has only a C-flanking codon and is therefore an N-terminal extension.\")\n",
    "        instype=\"N_ext\"\n",
    "    elif i_Nflank and not i_Cflank:\n",
    "        print(\"Insertion has only an N-flanking codon and is therefore an N-terminal extension.\")\n",
    "        instype=\"C_ext\"\n",
    "        \n",
    "    print(\"The insertion type interpereted is {}\".format(instype))\n",
    "    \n",
    "    #All indeces have been gathered. The indeces will be passed to the record_insertion_info() function to record the insertion in the file\n",
    "    record_insertion_info(j,instype,i_start,i_end,alignment,reference,ref_index,outfile,i_Nflank,i_Cflank)\n",
    "    #Recording complete. Pass index of the C-flanking codon back to msa_iterator() if it exists \n",
    "    return (i_Cflank if i_Cflank else len(reference))\n",
    "\n",
    "def delins_case(j,i_start,alignment,reference,ref_index,outfile):\n",
    "    \"\"\"\n",
    "    insertion_case() will gather the following information regarding the insertion-deletion (indel) event in 5 steps:\n",
    "    \n",
    "    i_start: The beginning of the insertion region. This is passed to the function by the function that triggered the delins case.\n",
    "    \n",
    "    i_end: The index of the end of the indel region (last location where the reference is blank and the variant genome has a codon)\n",
    "    \n",
    "    i_first: The index of the first codon deleted (used for computation of the variant code)\n",
    "    \n",
    "    i_last: The index of the last codon deleted (used for computation of the variant code)\n",
    "    \n",
    "    codons_inserted: The amino acid codons that have been inserted into the deletion region, with any blanks between residues removed.  \n",
    "    \"\"\"\n",
    "    #Step 1: Iterate until the C-terminal codon of the delins region is found\n",
    "    #This will be when there is an extant codon on both the reference strain and on genome j.\n",
    "    i=i_start\n",
    "    print(\"Iterating forward until the end the delins region is reached.\")\n",
    "    while True:\n",
    "        #Test the current codon.\n",
    "        print(\"Testing i={}\".format(i))\n",
    "        #If the current codon is an insertion, deletion, or a double blank, continue iteration.\n",
    "        if (reference[i]!=\"-\" and alignment[j,i]==\"-\") or (reference[i]==\"-\" and alignment[j,i]!=\"-\") or (reference[i]==\"-\" and alignment[j,i]==\"-\"):\n",
    "            print(\"The current codon is within the indel region. Continuing.\")\n",
    "            #Special case: the end of the genome is reached before a C-terminal flank is found\n",
    "            if i==len(reference)-1:\n",
    "                print(\"Reached the end of the loop without finding a C-terminal flank.\")\n",
    "                i_Cflank=None\n",
    "                #Break the while loop and continue to step 2\n",
    "                break\n",
    "            else:\n",
    "                i+=1\n",
    "                \n",
    "        #If there are codons on the reference and alignment genomes at the current position, the end of the indel region has been reached.\n",
    "        elif (reference[i]!=\"-\" and alignment[j,i]!=\"-\"):\n",
    "            print(\"Reached the C-terminal flanking codon to the indel region.\")\n",
    "            #Store this position as the position of the C-terminal flanking codon to the indel region (iteration in msa_iterator() must resume from this point.) \n",
    "            i_Cflank=i\n",
    "            break #Break the while loop and continue to step 2.\n",
    "            \n",
    "    #Step 2: Iterate backward over any blanks that may exist until the last codon that is either inserted or deleted\n",
    "    #Record this position as i_end\n",
    "    print(\"Iterating backward to the end of the indel region.\")\n",
    "    for i in range((i_Cflank if i_Cflank else len(reference)-1),i_start-1,-1):\n",
    "        print(\"Testing i={}\".format(i))\n",
    "        #If a blank is discovered at the position, continue to previous codon.\n",
    "        if (reference[i]==\"-\" and alignment[j,i]==\"-\"):\n",
    "            print(\"Skipping backwards over blank.\")\n",
    "            continue\n",
    "        \n",
    "        #If a deletion or an insertion is found, stop iteration and record current position as i_end (end of indel region).\n",
    "        elif (reference[i]!=\"-\" and alignment[j,i]==\"-\") or (reference[i]==\"-\" and alignment[j,i]!=\"-\"):\n",
    "            print(\"Discovered end of indel region. Recording index.\")\n",
    "            i_end=i\n",
    "            break\n",
    "    \n",
    "    #Step 3: Iterate through the indel region on the reference strand, storing the indeces of the first and last codons\n",
    "    #The first and last codons on the reference strand are the first and last amino acids deleted (necessary for code).\n",
    "    #Iteration is through region from i_start up to and including i_end.\n",
    "    \n",
    "    #Step 3.1: Finding the first codon\n",
    "    for i in range(i_start,i_end+1):\n",
    "        if (reference[i]==\"-\"):\n",
    "            continue\n",
    "        elif (reference[i]!=\"-\"):\n",
    "            print(\"Found the first codon of the reference strand (i={}).\".format(i))\n",
    "            #Store the index of the first codon\n",
    "            i_first=i\n",
    "            break\n",
    "    \n",
    "    #Step 3.2: Finding the last codon\n",
    "    for i in range(i_end,i_start-1,-1):\n",
    "        if (reference[i]==\"-\"):\n",
    "            continue\n",
    "        elif (reference[i]!=\"-\"):\n",
    "            print(\"Found the last codon of the reference strand (i={}).\".format(i))\n",
    "            #Store the index of the last codon\n",
    "            i_last=i\n",
    "            break\n",
    "            \n",
    "    #Step 4: Iterate through the slice of the variant strand and store the codons, removing any blanks \n",
    "    #The function insertion_var_AA() will do this.\n",
    "    codons_inserted=AA_blank_remover(i_start,i_end,j,alignment)\n",
    "    \n",
    "    #Step 5: Record information using the indeces discovered.\n",
    "    record_delins_info(i_start,i_end,i_first,i_last,j,codons_inserted,alignment,reference,ref_index,outfile)\n",
    "    \n",
    "    #Must return the position of the C-flanking codon to the indel region, so msa_iterator() does not record the region more than once.\n",
    "    return (i_Cflank if i_Cflank else len(reference)) #If the end of the genome is reached, pass that index to msa_iterator().\n",
    "    \n",
    "def record_deletion_info(j,i_start,i_end,alignment,reference,ref_index,outfile):\n",
    "    \"\"\"\n",
    "    Information recording\n",
    "    \"\"\"\n",
    "    #Store the name of the cluster associated with the variant sequence (sequence index j)\n",
    "    #The cluster name is in the id of the RefSeq object, before the \";\".\n",
    "    clustname=alignment[j].id.split(sep=\";\")[0]\n",
    "    #Retrieve size of cluster associated with the variant sequence j\n",
    "    #The cluster size is printed betweeen \";\" and \"=\" in id\n",
    "    clustsize=alignment[j].id.split(sep=\"=\")[1].split(sep=\";\")[0]\n",
    "\n",
    "    #Conditional: for single codon deletions, positions and amino acid residues will be stored as a single entry\n",
    "    #For multi-codon deletions, data will be stored as a range\n",
    "    #For single-codon deletions\n",
    "    if i_start==i_end:\n",
    "        #Store position of variant in MSA\n",
    "        msa_start=i_end+1\n",
    "        msa_end=\"-\"\n",
    "        #Store position of variant in reference genome\n",
    "        ref_start=ref_index[i_end]\n",
    "        ref_end=\"-\"\n",
    "        #Store codon in position i of reference genome\n",
    "        ref_AA=reference[i_end]\n",
    "        #Store the variant codon\n",
    "        var_AA=\"-\" #Retrieves the codon from the sequence at position j, at position i    \n",
    "        #Store the deletion code\n",
    "        \"\"\"\n",
    "        Based on the sequence variant nomenclature suggested by the Human Genome Variation Society,\n",
    "        the code format is <Amino_Acid_Deleted><Position_of_Deleted_Codon, reference Genome>del\n",
    "        source: (https://varnomen.hgvs.org/recommendations/protein/variant/deletion/)\n",
    "        \"\"\"\n",
    "        code=\"{}{}del\".format(ref_AA,ref_start)\n",
    "    #For multi-codon deletions\n",
    "    else:\n",
    "        #Store position range in MSA\n",
    "        msa_start=i_start+1\n",
    "        msa_end=i_end+1\n",
    "\n",
    "        #Store position range in reference genome\n",
    "        ref_start=ref_index[i_start]\n",
    "        ref_end=ref_index[i_end]\n",
    "\n",
    "        #Store reference amino acids that have been deleted in sequence j\n",
    "        #Slicing does not include the end value, so the end value should be one index after the end of the deletion region (i)\n",
    "        ref_AA=deletion_ref_AA(i_start,i_end,reference)\n",
    "\n",
    "        print(ref_AA)\n",
    "        #Store a blank for deleted variant codons\n",
    "        var_AA=\"-\"\n",
    "        print(var_AA)\n",
    "        #Store the deletion code\n",
    "        \"\"\"\n",
    "        Based on the sequence variant nomenclature suggested by the Human Genome Variation Society,\n",
    "        the code format is <Beginning_Amino_Acid><Beginning_Codon_Position, reference Genome>_<Ending_Amino_Acid><Ending_Codon_Position>del\n",
    "        source: (https://varnomen.hgvs.org/recommendations/protein/variant/deletion/)\n",
    "        \"\"\"\n",
    "        code=\"{}{}_{}{}del\".format(reference[i_start],ref_index[i_start],reference[i_end],ref_index[i_end])\n",
    "    \n",
    "    #Write above information to file\n",
    "    write_line(outfile,clustname,clustsize,\"del\",code,ref_AA,var_AA,msa_start,ref_start,msa_end,ref_end)\n",
    "    \n",
    "def record_insertion_info(j,instype,i_start,i_end,alignment,reference,ref_index,outfile,i_Nflank=None,i_Cflank=None):\n",
    "    #Store the name of the cluster associated with the variant sequence (sequence index j)\n",
    "    #The cluster name is in the id of the RefSeq object, before the \";\".\n",
    "    clustname=alignment[j].id.split(sep=\";\")[0]\n",
    "    #Retrieve size of cluster associated with the variant sequence j\n",
    "    #The cluster size is printed betweeen \";\" and \"=\" in id\n",
    "    clustsize=alignment[j].id.split(sep=\"=\")[1].split(sep=\";\")[0]\n",
    "    \n",
    "    #Store starting position on the multiple sequence alignment\n",
    "    msa_start=i_start+1\n",
    "    ref_AA=\"-\"\n",
    "    \n",
    "    #msa_end and var_AA: depend on whether the insertion is a single base or a multi-base insertion \n",
    "    if i_start==i_end:\n",
    "        msa_end=\"-\"\n",
    "        var_AA=alignment[j,i_start]\n",
    "    else:\n",
    "        msa_end=i_end+1\n",
    "        var_AA=AA_blank_remover(i_start,i_end,j,alignment)\n",
    "    \n",
    "    #ref_start and ref_end: defined as the reference positions of the N- and C- terminal flanking codons if they exist\n",
    "    ref_end=(ref_index[i_Cflank] if i_Cflank else \"-\") #C-flank is the flank downstream of the insertion.\n",
    "    ref_start=(ref_index[i_Nflank] if i_Nflank else \"-\")\n",
    "    \n",
    "    #The flanking reference residues are stored for creation of the code, if they exist.\n",
    "    Nflank=(reference[i_Nflank] if i_Nflank else None)\n",
    "    Cflank=(reference[i_Cflank] if i_Cflank else None)\n",
    " \n",
    "    \n",
    "    #The \"type\" and code fields depend on whether the insertion is a normal case, an N-terminal extension, or a C-terminal extension.\n",
    "    #TODO: instype will be passed through for output. Delete this later.\n",
    "    if instype==\"ins\":\n",
    "        print_type=instype\n",
    "        #Code is formatted as <N_flank amino acid ><>_<><>ins<var_AA>\n",
    "        code=\"{}{}_{}{}ins{}\".format(Nflank,ref_start,Cflank,ref_end,var_AA)\n",
    "        \n",
    "    elif instype==\"N_ext\":\n",
    "        #print type stores the type to be printed to the output file, as well as the type indicated in the code\n",
    "        #Both N-terminal and C-terminal extentions will be represented as \"ext\"\n",
    "        print_type=\"ext\"\n",
    "        #Code: will be formatted <C-flanking amino acid>1ext<Var_AA inserted>\n",
    "        #TODO: This is at odds with the Human Genome Variation Society, which reccomends that extentions describe the number of new codons instead of their type.\n",
    "        #I chose to print the type because we don't have the genetic information in this experiment. Should I keep things this way or change the format to match the reccomendations?\n",
    "        code=\"{}1ext{}\".format(Cflank,var_AA)\n",
    "        \n",
    "    elif instype==\"C_ext\":\n",
    "        print_type=\"ext\"\n",
    "        #Code: will be formatted <N-flanking amino acid><reference position of N-flanking amino acid>ext<Var_AA inserted>\n",
    "        code=\"{}{}ext{}\".format(Nflank,ref_start,var_AA)\n",
    "    \n",
    "    #Write infomation to file \n",
    "    write_line(outfile,clustname,clustsize,print_type,code,ref_AA,var_AA,msa_start,ref_start,msa_end,ref_end)\n",
    "    \n",
    "def record_delins_info(i_start,i_end,i_first,i_last,j,codons_inserted,alignment,reference,ref_index,outfile):\n",
    "    ###Record cluster information###\n",
    "    #Store the name of the cluster associated with the variant sequence (sequence index j)\n",
    "    #The cluster name is in the id of the RefSeq object, before the \";\".\n",
    "    clustname=alignment[j].id.split(sep=\";\")[0]\n",
    "    #Retrieve size of cluster associated with the variant sequence j\n",
    "    #The cluster size is printed betweeen \";\" and \"=\" in id\n",
    "    clustsize=alignment[j].id.split(sep=\"=\")[1].split(sep=\";\")[0]\n",
    "    \n",
    "    #Record type (always equal to delins)\n",
    "    print_type=\"delins\"\n",
    "    \n",
    "    ###Record indel code###\n",
    "    #Format: <first codon deleted><position of first codon deleted>_<last codon deleted><position of last codon deeted>delins<AA codons inserted>\n",
    "    \n",
    "    #Store identity and position of first codon deleted, using its index\n",
    "    del_AA_first=reference[i_first]\n",
    "    #ref_start=Position of first codon deleted\n",
    "    ref_start=ref_index[i_first]\n",
    "    \n",
    "    #Code format depends on if there is one or more codons deleted. \n",
    "    #If multiple codons are deleted (i_first!=i_last), calculate the code in the format below.\n",
    "    if i_first!=i_last:\n",
    "        #Store identity and position of last codon deleted, using its index\n",
    "        del_AA_last=reference[i_last]\n",
    "        #ref_end=Position of last codon deleted\n",
    "        ref_end=ref_index[i_last]\n",
    "        #Indel code\n",
    "        code=\"{}{}_{}{}delins{}\".format(del_AA_first,ref_start,del_AA_last,ref_end,codons_inserted)\n",
    "    #Follow the conditional below for single-codon deletions.\n",
    "    else:\n",
    "        ref_end=\"-\" #will store a blank on the raw data output file.\n",
    "        code=\"{}{}delins{}\".format(del_AA_first,ref_start,codons_inserted)\n",
    "        \n",
    "    ###Record AA sequence on reference and variant strands###\n",
    "    #ref_AA and var_AA will simply be the reference sequence and the variant sequence in the indel region,\n",
    "    #Without blanks removed.\n",
    "    ref_AA=str(reference[i_start:i_end+1])\n",
    "    var_AA=str(alignment[j,i_start:i_end+1].seq)\n",
    "    \n",
    "    #Record start and end positions on multiple sequence alignment (equal to index of start and end of region plus one)\n",
    "    msa_start=i_start+1\n",
    "    msa_end=i_end+1\n",
    "    \n",
    "    #Write information to file\n",
    "    write_line(outfile,clustname,clustsize,print_type,code,ref_AA,var_AA,msa_start,ref_start,msa_end,ref_end)\n",
    "    \n",
    "def deletion_ref_AA(i_start,i_end,reference):\n",
    "    \"\"\"\n",
    "    For multi-base deletions, records the refernce AA's deleted without blanks that may appear in the aligned reference genome.\n",
    "    \n",
    "    This function should only be called on multi-base deletions, otherwise an error will occur\n",
    "    \"\"\"\n",
    "    ref_AA=reference[i_start]\n",
    "    #the amino acid at i_start is included first, and iteration goes up to and including i_end (must be i_end+1 because range() does not include stop value)\n",
    "    for i in range(i_start+1,i_end+1):\n",
    "        #If an amino acid exists on the reference genome (no blank on msa), add the AA to the sequence\n",
    "        if reference[i]!=\"-\":\n",
    "            ref_AA=ref_AA+reference[i]\n",
    "    return ref_AA\n",
    "\n",
    "def AA_blank_remover(i_start,i_end,j,alignment):\n",
    "    \"\"\"\n",
    "    AA_blank_remover() records the amino acids within a specified region on a specified genome j, \n",
    "    excluding blanks that may exist due to the multiple sequence alignment.\n",
    "    \n",
    "    For example, an insertion reading \"FLP--AV\" in the multiple sequence alginment will be converted to \"FLPAV\".\n",
    "    \n",
    "    i_start: the starting index of the region (integer)\n",
    "    i_end: the ending index of the region (integer)\n",
    "    j: the index of the genome to be read (integer)\n",
    "    alingment: a biopython multiple sequence alignent object consisting of biopython SeqRecord objects.\n",
    "    \"\"\"\n",
    "    #Begin with an empty variable \n",
    "    AA_Seq=None\n",
    "    #Iterate from the first codon in the region to the last, including both first and last\n",
    "    for i in range(i_start,i_end+1):\n",
    "        #Blanks will be skipped by the conditionals below.\n",
    "        #The first non-blank character will be detected and assigned to AA_seq\n",
    "        if alignment[j,i]!=\"-\" and not AA_Seq:\n",
    "            AA_Seq=alignment[j,i]\n",
    "        #If there is at least one codon already stored in AA_seq, the next codon will be appended to the existing sequence.\n",
    "        elif alignment[j,i]!=\"-\" and AA_Seq:\n",
    "            AA_Seq=AA_Seq+alignment[j,i]\n",
    "    return AA_Seq\n",
    "                \n",
    "\"\"\"\n",
    "Part 2: Functions for analysis of raw variant data and report production\n",
    "\"\"\"\n",
    "def initialize_report_files(outname):\n",
    "    \"\"\"\n",
    "    outname = the directory and file prefix of the output files. Specific files will contain \n",
    "    \n",
    "    intitialize_report_files will create four files containing reports on the mutations present.\n",
    "    subs_report,ins_report,and dels_report will contain files\n",
    "    \"\"\"\n",
    "    subs_report=outname+\"_substitutions.txt\"\n",
    "    ins_report=outname+\"_insertions.txt\"\n",
    "    dels_report=outname+\"_deletions.txt\"\n",
    "    delins_report=outname+\"_indels.txt\"\n",
    "    clusters_report=outname+\"_clusters.txt\"\n",
    "    \n",
    "    #Create blank files to be filled with information in later steps \n",
    "    for file in [subs_report,ins_report,dels_report,delins_report,clusters_report]:\n",
    "        with (open(file,\"w\")) as outfile:\n",
    "                outfile.write(\"\")\n",
    "    return subs_report,ins_report,dels_report,delins_report,clusters_report\n",
    "\n",
    "#Create a class for storing all of the fields in the raw data file created in part 1\n",
    "class VariantRecord:\n",
    "    \"\"\"\n",
    "    Creates a variant_record object that stores the data for each variant produced in part 1. \n",
    "    \n",
    "    --Instance variables--\n",
    "    clustname: Cluster_ID\n",
    "    clustsize: Number of sequences in cluster \n",
    "    print_type: Type of variant (insertion, deletion, substitution)\n",
    "    code: Mutation code (formatted based on the nomenclature indicated by the Human Genome Variation Society https://varnomen.hgvs.org/recommendations/protein/)\n",
    "    ref_AA: Amino Acid(s) of Reference Strand \n",
    "        For insertions, this is the two amino acid residues the insertion ocurrs between.\n",
    "        For deletions, this will be the reference residue(s) deleted \n",
    "    var_AA: Variant Amino Acid Residue\n",
    "        For multi-codon insertions, this field will contain multiple columns\n",
    "        For deletions, this will be blank\n",
    "    msa_start: Start position of variant on multiple sequence alignment\n",
    "        For insertions this will be the codon number on the multiple sequence alignment before the inserted codon(s)\n",
    "    msa_end: End position of variant on multiple sequence alignment \n",
    "        For insertions this will be the codon number on the multiple sequence alignment before the inserted codon(s)\n",
    "        For substitutions and single-base deletions this will be blank\n",
    "    ref_start: Start position of variant on reference strand \n",
    "        For insertions this will be the codon number on the reference strand before the inserted codon(s)\n",
    "    ref_end: End position of variant on reference strand\n",
    "        For insertions this will be the codon number on the reference strand before the inserted codon(s)\n",
    "        For substitutions and single-base deletions this will be blank\n",
    "    \"\"\"\n",
    "    def __init__(self,clustname,clustsize,print_type,code,ref_AA,var_AA,msa_start,ref_start,msa_end,ref_end):\n",
    "        self.__clustname=clustname\n",
    "        self.__clustsize=(int(clustsize) if clustsize!=\"-\" else None) \n",
    "        self.__print_type=print_type\n",
    "        self.__code=code\n",
    "        self.__ref_AA=ref_AA\n",
    "        self.__var_AA=var_AA\n",
    "        self.__msa_start=(int(msa_start) if msa_start!=\"-\" else None) #Blanks may exist in this region in the output file\n",
    "        self.__msa_end=(int(msa_end) if msa_end!=\"-\" else None)\n",
    "        self.__ref_start=(int(ref_start) if ref_start!=\"-\" else None)\n",
    "        self.__ref_end=(int(ref_end) if ref_end!=\"-\" else None)\n",
    "    #Functions to print each variable    \n",
    "    def get_clustname(self):\n",
    "        return self.__clustname\n",
    "    def get_clustsize(self):\n",
    "        return self.__clustsize\n",
    "    def get_print_type(self):\n",
    "        return self.__print_type\n",
    "    def get_code(self):\n",
    "        return self.__code\n",
    "    def get_ref_AA(self):\n",
    "        return self.__ref_AA\n",
    "    def get_var_AA(self):\n",
    "        return self.__var_AA\n",
    "    def get_msa_start(self):\n",
    "        return self.__msa_start\n",
    "    def get_msa_end(self):\n",
    "        return self.__msa_end\n",
    "    def get_ref_start(self):\n",
    "        return self.__ref_start\n",
    "    def get_ref_end(self):\n",
    "        return self.__ref_end\n",
    "    #Define output upon calling print() on the VariantRecord object\n",
    "    def __str__(self):\n",
    "        return \"[{:<25}{:<5}{:<8}{:<15}]\".format(self.__clustname,self.__clustsize,self.__print_type,self.__code)\n",
    "        \n",
    "def initialize_mutation_database(outname):\n",
    "    \"\"\"\n",
    "    initialize_mutation_database will create a list of VariantRecord objects using the raw data file produced from \n",
    "    iteration through the multiple sequence alignment.\n",
    "    \"\"\"\n",
    "    var_database=[]\n",
    "    with open(outname+\"_variants_raw.tsv\",\"r\") as file:\n",
    "        #Load lines\n",
    "        lines=file.readlines()\n",
    "        #Exclude header\n",
    "        for line in lines[1:]:\n",
    "            #Remove newline characters \n",
    "            line=line.strip()\n",
    "            #Store each field in its corresponding VariantRecord attribute\n",
    "            clustname,clustsize,print_type,code,ref_AA,var_AA,msa_start,ref_start,msa_end,ref_end = line.split(\"\\t\")\n",
    "            #Use attributes from the line to create a VariantRecord object, and add the object to the list.\n",
    "            var_database.append(VariantRecord(clustname,clustsize,print_type,code,ref_AA,var_AA,msa_start,ref_start,msa_end,ref_end))\n",
    "    return var_database    \n",
    "\n",
    "#natsorted package used in the function below\n",
    "from natsort import natsorted\n",
    "\n",
    "def info_by_cluster(var_database,clusters_report):\n",
    "    \"\"\"\n",
    "    info_by_cluster() will produce an output file listing the mutations present in each cluster, \n",
    "    sorted in ascending order, alphabetically and then numerically, by cluster ID.\n",
    "    \n",
    "    var_database: a list of VariantRecord objects created from the intermediate tab-separated file \n",
    "        produced by msa_iterator().\n",
    "    \n",
    "    clusters_report: the path of the file name for exporting info by cluster, created by \n",
    "        initialize_report_files().\n",
    "    \"\"\"\n",
    "    #Get a list of unique cluster IDs\n",
    "    #A set is used to retrieve the unique cluster IDs (sets store only unique values, with no ordering)\n",
    "    clust_ids=set()\n",
    "    for Record in var_database:\n",
    "        clust_ids.add(Record.get_clustname())\n",
    "    #Convert set back to a list to allow sorting\n",
    "    clust_ids=list(clust_ids)\n",
    "    #Sort based on ascending cluster ID number (requires natural sorting from the natsort package)\n",
    "    clust_ids=natsorted(clust_ids)\n",
    "    \n",
    "    #Print info to outname_clusters.txt\n",
    "    with open(clusters_report,\"a\") as file:\n",
    "        #Print header to file\n",
    "        print(\"-\"*30,\"Variants observed by cluster\",\"-\"*30,sep=\"\\n\",end=\"\\n\",file=file)\n",
    "\n",
    "        #Iterate through unique cluster IDs, printing all variants associated with each cluster ID\n",
    "        for clustname in clust_ids:\n",
    "            #Iteratively subset database for the cluster id\n",
    "            subset=[record for record in var_database if record.get_clustname()==clustname]\n",
    "            #Print the header for the cluster id, then the number of sequences, then list the mutation code of each mutation event observed.\n",
    "            #End the print statement with a double new line to leave one line of blank space between each cluster in the record.\n",
    "            print(\">{}, {} sequences\".format(clustname,subset[0].get_clustsize()),*(record.get_code() for record in subset),sep=\"\\n\",end=\"\\n\\n\",file=file)\n",
    "\n",
    "def info_by_code(var_database,var_type,subs_report,ins_report,dels_report,delins_report):\n",
    "    \"\"\"\n",
    "    info_by_code() will return information on the identity of clusters with a variant, for each unique variant \n",
    "    in the multiple sequence alignment. \n",
    "    \n",
    "    var_database: A list of VariantRecord objects generated from the intermediate tab-separated file produced by msa_iterator().\n",
    "    \n",
    "    var_type: The type of variants to return information for. Accepts \"sub\" (substitutions),\"ins\" (insertions),\n",
    "        \"del\"(deletions), and \"delins\"(insertion-deletions).\n",
    "    \n",
    "    subs_report: the path of the file name for exporting substitution info by code, created by \n",
    "        initialize_report_files().\n",
    "        \n",
    "    ins_report: the path of the file name for exporting insertion info by code, created by \n",
    "        initialize_report_files().\n",
    "    \n",
    "    dels_report: the path of the file name for exporting deletion info by code, created by \n",
    "        initialize_report_files().\n",
    "        \n",
    "    delins_report: the path of the file name for exporting insertion-deletion info by code, created by \n",
    "        initialize_report_files().\n",
    "    \"\"\"\n",
    "    if var_type==\"sub\":\n",
    "        #subset for substitutions\n",
    "        substitutions=[record for record in var_database if record.get_print_type()==\"sub\"]\n",
    "\n",
    "        #Create a set of unique mutation codes\n",
    "        #Sorting should be performed based on the start position on the multiple sequence alignment. A dictionary is used to sort this\n",
    "        subcodes={}\n",
    "        for record in substitutions:\n",
    "            #Conditional below returns a dictionary of unique mutation codes\n",
    "            if record.get_code() not in subcodes:\n",
    "                subcodes[record.get_code()]=record.get_msa_start()\n",
    "        #Sort the dictionary by value of MSA start position (returns a list of the substitution codes sorted by msa start position)\n",
    "        subcodes=sorted(subcodes, key=lambda x:subcodes[x])\n",
    "    \n",
    "        #Print info to outname_substitutions.txt\n",
    "        with open(subs_report,\"a\") as file:\n",
    "            #Print header to file\n",
    "            print(\"-\"*40,\"Clusters by Mutation Code: Substitutions\",\"-\"*40,sep=\"\\n\",end=\"\\n\",file=file)\n",
    "\n",
    "            #For each unique code, subset the substitutions database by that code and print statistics\n",
    "            for code in subcodes:\n",
    "                #Iteratively subset database for the mutation code\n",
    "                subset=[record for record in var_database if record.get_code()==code]\n",
    "                #Calculate total number of sequences with the mutation\n",
    "                total_seq=0\n",
    "                for record in subset:\n",
    "                    #For each cluster with the mutation code, get the size of the cluster and add that to the total number of sequences\n",
    "                    total_seq+=record.get_clustsize()\n",
    "                #Print the header for the mutation code, then the total number of sequences with the code\n",
    "                #Next, the cluster name and number of sequences in the cluster is printed for each cluster with the code \n",
    "                #End the print statement with a double new line to leave one line of blank space between each cluster in the record.\n",
    "                print(\">{}, {} sequences\".format(code,total_seq),*(\"{}; {} sequences\".format(record.get_clustname(),record.get_clustsize()) for record in subset),sep=\"\\n\",end=\"\\n\\n\",file=file) \n",
    "\n",
    "    elif var_type==\"ins\":\n",
    "        #Subset for insertions\n",
    "        insertions=[record for record in var_database if record.get_print_type()==\"ext\" or record.get_print_type()==\"ins\"]\n",
    "        #Create a set of unique mutation codes\n",
    "        #Sorting should be performed based on the start position on the multiple sequence alignment. A dictionary is used to sort this\n",
    "        inscodes={}\n",
    "        for record in insertions:\n",
    "            #Conditional below returns a dictionary of unique mutation codes\n",
    "            if record.get_code() not in inscodes:\n",
    "                inscodes[record.get_code()]=record.get_msa_start()\n",
    "        #Sort the dictionary by value of MSA start position (returns a list of the insertion codes sorted by msa start position)\n",
    "        inscodes=sorted(inscodes, key=lambda x:inscodes[x])\n",
    "        \n",
    "        #Print information to outname_insertions.txt\n",
    "        with open(ins_report,\"a\") as file:\n",
    "            #Print header to file\n",
    "            print(\"-\"*40,\"Clusters by Mutation Code: Insertions\",\"-\"*40,sep=\"\\n\",end=\"\\n\",file=file)\n",
    "            \n",
    "            #For each unique code, subset the insertions database by that code and print statistics            \n",
    "            for code in inscodes:\n",
    "                #Iteratively subset database for the mutation code\n",
    "                subset=[record for record in var_database if record.get_code()==code]\n",
    "                #Calculate total number of sequences with the mutation\n",
    "                total_seq=0\n",
    "                for record in subset:\n",
    "                    #For each cluster with the mutation code, get the size of the cluster and add that to the total number of sequences\n",
    "                    total_seq+=record.get_clustsize()\n",
    "                #Print the header for the mutation code, then the total number of sequences with the code\n",
    "                #Next, the cluster name and number of sequences in the cluster is printed for each cluster with the code \n",
    "                #End the print statement with a double new line to leave one line of blank space between each cluster in the record.\n",
    "                print(\">{}, {} sequences\".format(code,total_seq),*(\"{}; {} sequences\".format(record.get_clustname(),record.get_clustsize()) for record in subset),sep=\"\\n\",end=\"\\n\\n\",file=file)\n",
    "\n",
    "    elif var_type==\"del\":\n",
    "        #Subset for deletions\n",
    "        dels=[record for record in var_database if record.get_print_type()==\"del\"]\n",
    "\n",
    "        #Create a set of unique mutation codes\n",
    "        #Sorting should be performed based on the start position on the multiple sequence alignment. A dictionary is used to sort this\n",
    "        delcodes={}\n",
    "        for record in dels:\n",
    "            #Conditional below returns a dictionary of unique mutation codes\n",
    "            if record.get_code() not in delcodes:\n",
    "                delcodes[record.get_code()]=record.get_msa_start()\n",
    "        #Sort the dictionary by value of MSA start position (returns a list of the deletion codes sorted by msa start position)\n",
    "        delcodes=sorted(delcodes, key=lambda x:delcodes[x])\n",
    "\n",
    "        with open(dels_report,\"a\") as file:\n",
    "            #Print header to file\n",
    "            print(\"-\"*40,\"Clusters by Mutation Code: Deletions\",\"-\"*40,sep=\"\\n\",end=\"\\n\",file=file)\n",
    "\n",
    "            #For each unique code, subset the substitutions database by that code and print statistics\n",
    "            for code in delcodes:\n",
    "                #Iteratively subset database for the mutation code\n",
    "                subset=[record for record in var_database if record.get_code()==code]\n",
    "                #Calculate total number of sequences with the mutation\n",
    "                total_seq=0\n",
    "                for record in subset:\n",
    "                    #For each cluster with the mutation code, get the size of the cluster and add that to the total number of sequences\n",
    "                    total_seq+=record.get_clustsize()\n",
    "                #Print the header for the mutation code, then the total number of sequences with the code\n",
    "                #Next, the cluster name and number of sequences in the cluster is printed for each cluster with the code \n",
    "                #End the print statement with a double new line to leave one line of blank space between each cluster in the record.\n",
    "                print(\">{}, {} sequences\".format(code,total_seq),*(\"{}; {} sequences\".format(record.get_clustname(),record.get_clustsize()) for record in subset),sep=\"\\n\",end=\"\\n\\n\",file=file) \n",
    "\n",
    "    elif var_type==\"indel\":\n",
    "        #Subset for indels\n",
    "        indels=[record for record in var_database if record.get_print_type()==\"delins\"]\n",
    "\n",
    "        #Create a set of unique mutation codes\n",
    "        #Sorting should be performed based on the start position on the multiple sequence alignment. A dictionary is used to sort this\n",
    "        indelcodes={}\n",
    "        for record in indels:\n",
    "            #Conditional below returns a dictionary of unique mutation codes\n",
    "            if record.get_code() not in indelcodes:\n",
    "                indelcodes[record.get_code()]=record.get_msa_start()\n",
    "        #Sort the dictionary by value of MSA start position (returns a list of the insertion-deletion codes sorted by msa start position)\n",
    "        indelcodes=sorted(indelcodes, key=lambda x:indelcodes[x])\n",
    "        \n",
    "        with open(delins_report,\"a\") as file:\n",
    "            #Print header to file\n",
    "            print(\"-\"*40,\"Clusters by Mutation Code: Indels\",\"-\"*40,sep=\"\\n\",end=\"\\n\",file=file)\n",
    "            #For each unique code, subset the substitutions database by that code and print statistics\n",
    "            for code in indelcodes:\n",
    "                #Iteratively subset database for the mutation code\n",
    "                subset=[record for record in var_database if record.get_code()==code]\n",
    "                #Calculate total number of sequences with the mutation\n",
    "                total_seq=0\n",
    "                for record in subset:\n",
    "                    #For each cluster with the mutation code, get the size of the cluster and add that to the total number of sequences\n",
    "                    total_seq+=record.get_clustsize()\n",
    "                #Print the header for the mutation code, then the total number of sequences with the code\n",
    "                #Next, the cluster name and number of sequences in the cluster is printed for each cluster with the code \n",
    "                #End the print statement with a double new line to leave one line of blank space between each cluster in the record.\n",
    "                print(\">{}, {} sequences\".format(code,total_seq),*(\"{}; {} sequences\".format(record.get_clustname(),record.get_clustsize()) for record in subset),sep=\"\\n\",end=\"\\n\\n\",file=file) \n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Program Execution"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "\"\"\"\n",
    "Fake test AA sequence to check function of program\n",
    "\"\"\"\n",
    "#Load MSA, store as biopython multiple sequence alingment object\n",
    "alignment = msa_in.read(\"msa_tester.fasta\",\"fasta\")\n",
    "\n",
    "#Initialize files\n",
    "#Define output file name\n",
    "outname = \"../Variant_Call/test_tester\"\n",
    "raw_out = initalize_file(outname)\n",
    "\n",
    "#Define reference genome (here it is index 1 of the multiple sequence alignment)\n",
    "reference=alignment[0].seq\n",
    "#Create reference sequence object storing reference sequence indeces corresponding to msa index (if \"-\" in reference seq after alignment)\n",
    "ref_index=reference_position(reference)\n",
    "\n",
    "#Iterate through multiple sequence alignment and write raw data from variant information to file\n",
    "msa_iterator1(alignment,raw_out,reference,ref_index,raw_out)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "\"\"\"\n",
    "Part 1: Generate raw variant data from multiple sequence alignment input\n",
    "\"\"\"\n",
    "#Load MSA, store as biopython multiple sequence alingment object\n",
    "alignment = msa_in.read(\"../Aligned_Sequences/spikeprot0902_full_aligned_nosingletons_1iter.fasta\",'fasta')\n",
    "\n",
    "#Initialize files\n",
    "#Define output file name\n",
    "outname = \"../Variant_Call/test\"\n",
    "raw_out = initalize_file(outname)\n",
    "\n",
    "#Define reference genome (here it is index 1 of the multiple sequence alignment)\n",
    "reference=alignment[1].seq\n",
    "#Create reference sequence object storing reference sequence indeces corresponding to msa index (if \"-\" in reference seq after alignment)\n",
    "ref_index=reference_position(reference)\n",
    "\n",
    "#Iterate through multiple sequence alignment and write raw data from variant information to file\n",
    "msa_iterator1(alignment,raw_out,reference,ref_index,raw_out)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Step 2: Report information from raw file\n",
    "outname = \"../Variant_Call/test_tester\"\n",
    "var_database=initialize_mutation_database(outname)\n",
    "\n",
    "#Initialize output files\n",
    "subs_report,ins_report,dels_report,delins_report,clusters_report=initialize_report_files(outname)\n",
    "\n",
    "#Print mutation events observed in each cluster\n",
    "info_by_cluster(var_database,clusters_report)\n",
    "\n",
    "#Print substitution events by code\n",
    "info_by_code(var_database,\"sub\",subs_report,ins_report,dels_report,delins_report)\n",
    "\n",
    "#Print insertion events by code\n",
    "info_by_code(var_database,\"ins\",subs_report,ins_report,dels_report,delins_report)\n",
    "\n",
    "#Print deletion events by code\n",
    "info_by_code(var_database,\"del\",subs_report,ins_report,dels_report,delins_report)\n",
    "\n",
    "#Print indel events by code\n",
    "info_by_code(var_database,\"indel\",subs_report,ins_report,dels_report,delins_report)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
